ports/rp2/machine_uart.c

```diff
@@ -477,7 +477,7 @@ STATIC mp_uint_t machine_uart_ioctl(mp_obj_t self_in, mp_uint_t request, mp_uint
     if (request == MP_STREAM_POLL) {
         uintptr_t flags = arg;
         ret = 0;
-        if ((flags & MP_STREAM_POLL_RD) && ringbuf_avail(&self->read_buffer) > 0) {
+        if ((flags & MP_STREAM_POLL_RD) && MP_OBJ_SMALL_INT_VALUE(machine_uart_any(self_in)) > 0) {
```

Comment:
would it be better (doesn't modify any state, is faster to execute) to use:
```
if ((flags & MP_STREAM_POLL_RD) && (uart_is_readable(self->uart) || ringbuf_avail(&self->read_buffer) > 0)) {
```
?

Reply:
@dpgeorge For some reason this is not the same. I have a nimble port for RP2 which uses machine_uart, and it's working fine with the suggested ioctl fix, but fails with the change above. I was going to send a PR later after this is merged, but maybe you can have a look and let me know why this doesn't work ?  #7561

Reply:
Interesting... try increasing the rxbuf length.  stm32 uses 768.

Reply:
No it doesn't help. I have an HCI controller driver in /drivers (which I'll also send later) I get zeros:
```
nina_bt_hci.c: unexpected HCI packet: 00
```

Reply:
Okay I think I know the issue, it's timing out too quickly because the default timeout is 0. When the FIFO is drained in ioctl instead read is always ready, so your suggestion is fine.

Reply:
Fixed.

Reply:
I think the default timeout should be 1 to allow it to check at least once after an ioctl, before timing out.

Reply:
Or, adjust the logic in `machine_uart_read` so that it checks `uart_is_readable` before returning EAGAIN.